File: mobile-android/app/src/test/java/ph/attendance/domain/ReportingTest.kt
package ph.attendance.domain

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class TallyTest {

    @Test
    fun `counts each status`() {
        val counts = listOf(
            record(status = AttendanceStatus.PRESENT, date = "2024-03-01"),
            record(status = AttendanceStatus.LATE, date = "2024-03-02"),
            record(status = AttendanceStatus.ABSENT, date = "2024-03-03"),
            record(status = AttendanceStatus.PRESENT, date = "2024-03-04"),
        ).tally()

        assertEquals(StatusTally(present = 2, late = 1, absent = 1), counts)
    }

    @Test
    fun `counts late as attending, not as an absence`() {
        // A student late every single day attended every single day. Folding late into absent
        // was a real reporting error in the original app.
        assertEquals(1f, StatusTally(present = 0, late = 10, absent = 0).rate, 0.0001f)
        assertEquals(0.8f, StatusTally(present = 5, late = 3, absent = 2).rate, 0.0001f)
    }

    @Test
    fun `reports zero rather than dividing by zero`() {
        assertEquals(0f, StatusTally().rate, 0.0001f)
    }

    @Test
    fun `formats as a whole percentage`() {
        assertEquals("80%", formatRate(0.8f))
        assertEquals("88%", formatRate(0.876f))
        assertEquals("0%", formatRate(0f))
    }
}

class SummariseStudentsTest {

    private val students = listOf(
        student(id = "stu-1", studentNumber = "s1", lastName = "Alvarez", firstName = "Ana"),
        student(id = "stu-2", studentNumber = "s2", lastName = "Bautista", firstName = "Ben"),
        student(id = "stu-3", studentNumber = "s3", lastName = "Castro", firstName = "Cara"),
    )

    private val records = listOf(
        record("s1", AttendanceStatus.PRESENT, "2024-03-01", id = "r1"),
        record("s1", AttendanceStatus.LATE, "2024-03-02", id = "r2"),
        record("s2", AttendanceStatus.ABSENT, "2024-03-01", id = "r3"),
        // Outside the range under test.
        record("s1", AttendanceStatus.PRESENT, "2024-02-28", id = "r4"),
        record("s2", AttendanceStatus.PRESENT, "2024-03-10", id = "r5"),
    )

    @Test
    fun `counts only records inside the range, inclusive of both ends`() {
        val byNumber = summariseStudents(students, records, "2024-03-01", "2024-03-05")
            .associateBy { it.student.studentNumber }

        assertEquals(StatusTally(present = 1, late = 1), byNumber.getValue("s1").counts)
        assertEquals(StatusTally(absent = 1), byNumber.getValue("s2").counts)
    }

    @Test
    fun `keeps students who have no records at all`() {
        // The student nobody ever scanned is precisely the one a coordinator is looking for.
        val cara = summariseStudents(students, records, "2024-03-01", "2024-03-05")
            .single { it.student.studentNumber == "s3" }

        assertEquals(StatusTally(), cara.counts)
        assertEquals(0, cara.sessions)
    }

    @Test
    fun `excludes archived students`() {
        val withArchived = students + student(id = "stu-4", studentNumber = "s4", archived = true)
        val summaries = summariseStudents(withArchived, records, "2024-03-01", "2024-03-05")
        assertTrue(summaries.none { it.student.studentNumber == "s4" })
    }

    @Test
    fun `sorts by display name`() {
        val summaries = summariseStudents(students, records, "2024-03-01", "2024-03-05")
        assertEquals(listOf("s1", "s2", "s3"), summaries.map { it.student.studentNumber })
    }
}

class DashboardTest {

    private val students = listOf(
        student(id = "stu-1", studentNumber = "s1", lastName = "Alvarez", gender = Gender.FEMALE),
        student(id = "stu-2", studentNumber = "s2", lastName = "Bautista", gender = Gender.MALE),
        student(id = "stu-3", studentNumber = "s3", lastName = "Castro", gender = Gender.MALE),
    )

    private val records = listOf(
        record("s1", AttendanceStatus.PRESENT, DATE, id = "r1"),
        record("s2", AttendanceStatus.LATE, DATE, id = "r2"),
        record("s1", AttendanceStatus.PRESENT, "2024-03-14", id = "r3"),
    )

    @Test
    fun `counts only the chosen day`() {
        val breakdown = buildDashboard(students, records, DATE)
        assertEquals(StatusTally(present = 1, late = 1), breakdown.counts)
    }

    @Test
    fun `lists students with no record as unaccounted for`() {
        val breakdown = buildDashboard(students, records, DATE)
        assertEquals(listOf("s3"), breakdown.unaccountedFor.map { it.studentNumber })
    }

    @Test
    fun `drops records whose student has since been removed`() {
        val orphaned = records + record("gone", AttendanceStatus.PRESENT, DATE, id = "r9")
        val breakdown = buildDashboard(students, orphaned, DATE)

        assertEquals(2, breakdown.entries.size)
        assertEquals(1, breakdown.counts.present)
    }

    @Test
    fun `splits by status and gender`() {
        val breakdown = buildDashboard(students, records, DATE)

        assertEquals(
            listOf("s1"),
            breakdown.entriesWith(AttendanceStatus.PRESENT).map { it.student.studentNumber },
        )
        assertEquals(
            listOf("s2"),
            breakdown.entriesWith(AttendanceStatus.LATE).ofGender(Gender.MALE)
                .map { it.student.studentNumber },
        )
        assertEquals(1, breakdown.entries.ofGender(Gender.FEMALE).size)
    }
}

File: web/tests/domain/reporting.test.ts
import { describe, expect, it } from "vitest";

import type { AttendanceRecord, AttendanceStatus } from "@/domain/model";
import {
  attendanceRate,
  buildDashboard,
  entriesByGender,
  entriesByStatus,
  formatRate,
  summariseStudents,
  tally,
} from "@/domain/reporting";

import { makeStudent } from "../fixtures";

function makeRecord(
  studentNumber: string,
  status: AttendanceStatus,
  date: string,
  id = 1,
): AttendanceRecord {
  return {
    id,
    scheduleId: 1,
    sectionId: 1,
    studentNumber,
    date,
    status,
    scheduleTitle: "Morning Assembly",
    recordedAt: `${date}T07:10:00.000Z`,
  };
}

describe("tally and attendanceRate", () => {
  it("counts each status", () => {
    const counts = tally([
      makeRecord("a", "present", "2024-03-01", 1),
      makeRecord("a", "late", "2024-03-02", 2),
      makeRecord("a", "absent", "2024-03-03", 3),
      makeRecord("a", "present", "2024-03-04", 4),
    ]);

    expect(counts).toEqual({ present: 2, late: 1, absent: 1 });
  });

  it("counts late as attending, not as an absence", () => {
    // A student who is late every single day attended every single day. Folding late into
    // absent here was a real reporting error in the original app.
    expect(attendanceRate({ present: 0, late: 10, absent: 0 })).toBe(1);
    expect(attendanceRate({ present: 5, late: 3, absent: 2 })).toBe(0.8);
  });

  it("reports zero rather than dividing by zero when nothing was recorded", () => {
    expect(attendanceRate({ present: 0, late: 0, absent: 0 })).toBe(0);
  });

  it("formats as a whole percentage", () => {
    expect(formatRate(0.8)).toBe("80%");
    expect(formatRate(0.876)).toBe("88%");
    expect(formatRate(0)).toBe("0%");
  });
});

describe("summariseStudents", () => {
  const students = [
    makeStudent({ id: 1, studentNumber: "s1", lastName: "Alvarez", firstName: "Ana" }),
    makeStudent({ id: 2, studentNumber: "s2", lastName: "Bautista", firstName: "Ben" }),
    makeStudent({ id: 3, studentNumber: "s3", lastName: "Castro", firstName: "Cara" }),
  ];

  const records = [
    makeRecord("s1", "present", "2024-03-01", 1),
    makeRecord("s1", "late", "2024-03-02", 2),
    makeRecord("s2", "absent", "2024-03-01", 3),
    // Outside the range under test.
    makeRecord("s1", "present", "2024-02-28", 4),
    makeRecord("s2", "present", "2024-03-10", 5),
  ];

  const range = { start: "2024-03-01", end: "2024-03-05" };

  it("counts only records inside the range, inclusive of both ends", () => {
    const summaries = summariseStudents(students, records, range);
    const byNumber = new Map(summaries.map((s) => [s.student.studentNumber, s]));

    expect(byNumber.get("s1")?.counts).toEqual({ present: 1, late: 1, absent: 0 });
    expect(byNumber.get("s2")?.counts).toEqual({ present: 0, late: 0, absent: 1 });
  });

  it("keeps students who have no records at all", () => {
    // The student nobody ever scanned is precisely the one a coordinator is looking for.
    const summaries = summariseStudents(students, records, range);
    const cara = summaries.find((s) => s.student.studentNumber === "s3");

    expect(cara).toBeDefined();
    expect(cara?.counts).toEqual({ present: 0, late: 0, absent: 0 });
    expect(cara?.sessions).toBe(0);
  });

  it("excludes archived students", () => {
    const withArchived = [...students, makeStudent({ id: 4, studentNumber: "s4", archived: true })];
    const summaries = summariseStudents(withArchived, records, range);

    expect(summaries.map((s) => s.student.studentNumber)).not.toContain("s4");
  });

  it("sorts by display name", () => {
    const summaries = summariseStudents(students, records, range);
    expect(summaries.map((s) => s.student.studentNumber)).toEqual(["s1", "s2", "s3"]);
  });
});

describe("buildDashboard", () => {
  const students = [
    makeStudent({ id: 1, studentNumber: "s1", lastName: "Alvarez", gender: "female" }),
    makeStudent({ id: 2, studentNumber: "s2", lastName: "Bautista", gender: "male" }),
    makeStudent({ id: 3, studentNumber: "s3", lastName: "Castro", gender: "male" }),
  ];

  const date = "2024-03-15";
  const records = [
    makeRecord("s1", "present", date, 1),
    makeRecord("s2", "late", date, 2),
    makeRecord("s1", "present", "2024-03-14", 3),
  ];

  it("counts only the chosen day", () => {
    const breakdown = buildDashboard(students, records, date);
    expect(breakdown.counts).toEqual({ present: 1, late: 1, absent: 0 });
  });

  it("lists students with no record at all as unaccounted for", () => {
    const breakdown = buildDashboard(students, records, date);
    expect(breakdown.unaccountedFor.map((s) => s.studentNumber)).toEqual(["s3"]);
  });

  it("drops records whose student has since been removed", () => {
    const orphaned = [...records, makeRecord("gone", "present", date, 9)];
    const breakdown = buildDashboard(students, orphaned, date);

    expect(breakdown.entries).toHaveLength(2);
    expect(breakdown.counts.present).toBe(1);
  });

  it("splits by status and gender", () => {
    const breakdown = buildDashboard(students, records, date);

    expect(entriesByStatus(breakdown, "present").map((e) => e.student.studentNumber)).toEqual([
      "s1",
    ]);
    expect(
      entriesByGender(entriesByStatus(breakdown, "late"), "male").map(
        (e) => e.student.studentNumber,
      ),
    ).toEqual(["s2"]);
    expect(entriesByGender(breakdown.entries, "female")).toHaveLength(1);
  });
});
